Tensor module: add matches methods for TensAdd, TensMul, and Tensor

Kishore96in opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

<!-- 
#### References to other Issues or PRs
-->
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
This PR adds `matches` routines for `TensMul` and `TensAdd`. This is needed in order to match subexpressions properly (similar to Add and Mul in sympy core).
I have also added a `matches` method for `Tensor` that takes index symmetries into account.

#### Other comments
The `matches` method I have implemented for `TensMul` only supports commutative tensors (will raise an error if it is used with non-commutative ones).

It is now possible to do things like

```

$isympy
In [1]: from sympy.tensor.tensor import TensorIndexType, tensor_indices, WildTensorInd
   ...: ex, TensorHead, WildTensorHead
In [2]: from sympy import Wild, Symbol
In [3]: R3 = TensorIndexType('R3', dim=3)
In [4]: r = tensor_indices("r", R3)
In [5]: wi = Wild("wi")
In [6]: a = symbols("a", cls = WildTensorIndex, tensor_index_type=R3, ignore_updown=Tr
   ...: ue)
In [7]: K = TensorHead("K", [R3])
In [8]: W = WildTensorHead('W', unordered_indices=True)
In [9]: k = Symbol("K")
In [10]: ( k**4*K(r)*K(-r) ).replace( wi*W()*K(a)*K(-a), wi*W()*k**2 ).doit()
Out[10]: 
 6
K
```

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* tensor
  * Implemented matching for TensAdd
  * Implemented matching for TensMul when its arguments are commutative.
  * Matching for Tensor now takes index symmetries into account.
<!-- END RELEASE NOTES -->

Kishore96in committed:
Add rudimentary matches method to TensMul

Kishore96in committed:
Remove a deubg statement

Kishore96in committed:
Remove debug statement

Kishore96in committed:
Raise NotImplementedError if tensors are non-commuting.

Kishore96in committed:
Proper implementation of matching for TensMul

Kishore96in committed:
Fix iterating over dictionary values

dict.values() is what I want, not dict.items()

Kishore96in committed:
Fix typo

Ugh.

Kishore96in committed:
Handle case where the Wild part has no free indices

Also add a doit to the TensMul call (otherwise, causes problems with
comparison later on)

Kishore96in committed:
Fix match condition for remaining_e_tensors

No need to try to match if nothing is remaining

Kishore96in committed:
Make a copy of repl_dict

Kishore96in committed:
Only add Wilds to repl_dict

Sympy does not like it when matches() returns non-Wild values.

Kishore96in committed:
Simplify and correct matching of remaining_e_tensors

Kishore96in committed:
Reject if coefficient does not match

Kishore96in committed:
Simplify code

Mostly renamed variables or introduced new ones

Kishore96in committed:
Replace internally contracted indices in query by dummies

I've run into a problem with Basic.match now, which is a bummer.

Kishore96in committed:
Split dummy-wild replacement into a separate function

Kishore96in committed:
wildify_dummies: only handles fellows which are not already wild.

Kishore96in committed:
Remove unnecessary check

Kishore96in committed:
TensMul: always replace dummies by WildTensorIndex

This allows one to simplify the matching code tremendously, I think.

Kishore96in committed:
Remove wildify_dummies

Not needed, now that TensMul automatically replaces dummies with
WildTensorIndex instances

Kishore96in committed:
Make sure the same tensor in the query is not mapped to multiple tensors in expr

Kishore96in committed:
Fix matching in case of contracted indices in the query.

Earlier, it could match, e.g., C_0 and C_0 in the query to different
tensors!

Kishore96in committed:
Fix matching of multiple WildTensors

Kishore96in committed:
Call doit() on TensMul

Kishore96in committed:
For indexless Wilds, use the .matches method

Does not change the behaviour, but should make it easier to
modify/extend the behavious of WildTensor

Kishore96in committed:
Handle the case where expr is a Tensor (not TensMul)

Sort of an edge case, but good that I've noticed it now.

Kishore96in committed:
Don't discard substitutions for Wild and WildFunction

Serves me right for doing it so hackily. Ugh. Need to clean things up.

Kishore96in committed:
Update docstring

Kishore96in committed:
Use repl_dict instead of creating a temporary dictionary

Kishore96in committed:
Copy repl_dict in the matches routine itself.

Kishore96in committed:
Handle absence of repl_dict in args of _matches_commutative

Had mistakenly removed it earlier, but it is still needed

Kishore96in committed:
Remove unused variables

query_wild_tensor_indices and free

All tests pass with these removed, so I guess it is indeed safe to
remove them.

Kishore96in committed:
Explanatory comments

Kishore96in committed:
Fix comment

Kishore96in committed:
Don't do unnecessary work when there are no Wilds

Huge speedup when doing an epsilon-delta replacement on a product of six
epsilons! (160s -> 3 s)

Kishore96in committed:
Move function and comment

Consolidate the 'matching simple patterns' part

Kishore96in committed:
Merge two if-else conditions

isinstance(self, TensExpr) is always True, since self is a TensMul

Kishore96in committed:
Fix stupid typo

Kishore96in committed:
Denest some loops

Not sifting the heads beforehand doesn't seem to incur a significant
performance penalty

Kishore96in committed:
Consistent nomenclature in comments

Kishore96in committed:
Consistent nomenclature

Kishore96in committed:
TensMul matching: some sanity checks and errors

Earlier, if we had an argument of unknown type, it was just being
ignored

Kishore96in committed:
Generalize sanity check

Kishore96in committed:
TensMul matching: handle Tensor and TensMul on an equal footing

Also simplify finding remaining tensors while matching indexless Wilds

Kishore96in committed:
Tensor: .matches method for tensor class

Kishore96in committed:
Simplify an if statement

Kishore96in committed:
TensMul: to match a tensor, just use its matches method

Earlier, the whole logic of the matches method was reproduces here.

Kishore96in committed:
tensor: Use doit instead of as_dummy

as_dummy does nothing for tensors right now.

Kishore96in committed:
TensMul matching: canonicalize while checking if same

Kishore96in committed:
TensMul: Move simple matching tests to a separate function

Kishore96in committed:
TensMul.match: handle case where expr.args contains a TensAdd

Kishore96in committed:
comment

Kishore96in committed:
else-if -> elif

Kishore96in committed:
matches method for TensAdd

Kishore96in committed:
Fix bug

Kishore96in committed:
Fix wild matching in TensAdd.matches

break should not be there, since I want to add all the matches for a
particular wild.

Kishore96in committed:
TensMul.matches: match indexless WildTensor with 1

Kishore96in committed:
TensMul.matches: add WildTensorHead to list of allowed wilds in repl_dict

Kishore96in committed:
TensMul._mathces_commutative: Pass the 'old' argument to the coeff matcher

Kishore96in committed:
Move matches_simple to TensExpr

I want to use it for TensAdd too (later)

Kishore96in committed:
TensMul._tensMul_contract_indices: calculate free, dummy data again after replacing indices

They can change due to the replacement in some edge cases.

Kishore96in committed:
Add tests for TensMul matching

Kishore96in committed:
TensMul: Don't unconditionally change dummy indices to WildTensorIndex

Just use the same type as the user specified.

Kishore96in committed:
TensMul._matches_commutative: fix index comparison

I want index equality checking (to find out which part of the TensMul
should be matched to the indexless wild) to ignore the distinction
between TensorIndex and WildTensorIndex

Kishore96in committed:
TensMul._matches_commutative: use .matches instead of a custom equality function for index comparison

Also removed that ugly _eq function

Kishore96in committed:
TensMul._matches_commutative: we now need to consider wild indices even if they are dummies!

Kishore96in committed:
TensMul._matches_commutative: Move _get_wilds out to global scope

Kishore96in committed:
TensAdd._matches: use _get_wilds instead of _get_nondummy_wilds

_get_nondummy_wilds was apparently removed by me much earlier. Seems
this codepath is not covered by the current tests, so I did not catch
it.

Kishore96in committed:
Test TensAdd.matches

Was untested before this.

Kishore96in committed:
Tests: split into smaller tests

Kishore96in committed:
test_tensor_matching: remove unnecessary newlines

Kishore96in committed:
test_TensMul_matching: remove unnecessary newlines

Kishore96in committed:
test_TensMul_matching: remove use of check_tens_eq

replace by assert

Kishore96in committed:
test_TensAdd_matching: remove use of check_tens_eq

replace by assert

Kishore96in committed:
test_TensAdd_matching: add a doit

Kishore96in committed:
test_TensMul_matching: consistent spacing

Kishore96in committed:
test_TensMul_matching: add an import

Kishore96in committed:
TensMul._matches_commutative: ensure that expr is a TensMul

Earlier, it would not reject things like PartialDerivative (which
inherit from TensExpr, but are not TensAdd)

Kishore96in committed:
TensMul._matches_commutative: try to use recursion

Tests pass, but code is not correct.

Kishore96in committed:
TensMul._matches_commutative: check internal consistency of index mapping

The two halves of a dummy pair should always be mapped to the same
index.

Kishore96in committed:
TensMul._matches_commutative: correctly pass rpl_dict during recursion.

Kishore96in committed:
test_TensMul_matching: add test for an edge case which was not caught earlier

This prompted the recent rewrite that uses recursion.

Kishore96in committed:
TensMul._matches_commutative: merge two loops

Kishore96in committed:
Tensor: add function _get_symmetrized_forms to permute tensor indices consistent with available symmetries

Kishore96in committed:
Tensor.matches: consider index symmetries

Kishore96in committed:
Tensor.matches: fix silly bug

Was returning None too early.

Kishore96in committed:
Add a test for tensor matching with symmetries

Kishore96in committed:
Tensor.matches: TODOs

Kishore96in committed:
TensMul._matches_commutative: explicitly consider symmetries of q_tensor.

Kinda hacky now, but eh.

Kishore96in committed:
test_TensMul_matching: add another test for tensor symmetry matching

Kishore96in committed:
Tensor.matches: remove resolved TODOs

Kishore96in committed:
Tensor._matches: remove TODO

Have made a note of it elsewhere

Kishore96in committed:
test_TensMul_matching: fix a test

Just had to add canon_bp since matching is now not deterministic.
In Tensor.matches, I internally use set(), and thus the form of the
results may change from invocation to invocation. Nevertheless, the
mathematical content of the results does not seem to change.

Kishore96in committed:
TensMul._matches_commutative: correctly find the non-tensor part of expr

This need not be the same as expr.coeff when expr.doit() has not been
called.

Kishore96in committed:
test_TensMul_matching: add test

Kishore96in committed:
TensMul._matches_commutative: account for the fact that a naked Expr may match a lone indexless WildTensor

Kishore96in committed:
TensMul._matches_commutative: shallow doit everywhere

deep doit seems like the wrong thing to do. All tests still pass after
this change.

Kishore96in committed:
tensor.py: _get_wilds: remove unnecessary list comprehension

Kishore96in committed:
TensExpr._matches_simple: try some cheaper checks before calling doit

doit can be quite expensive. Since _matches_simple is called often
during any match/replace operation on TensExpr, we expect this to
improve performance quite a bit.

Kishore96in committed:
TensExpr._matches_simple: TODO (avoiding doit)

Kishore96in committed:
TensExpr._matches_simple: exclude integrals from doit

integration is quite expensive, and anyway we mainly want doit here to
only relabel dummy indices, collect terms with common coeffs, and
flatten TensExpr.

Kishore96in committed:
TensExpr._matches_simple: remove call to doit()

I haven't run into a case where this is necessary, so let's assume it
isn't for now.

Kishore96in committed:
tensor: remove duplicated import of itertools

ThePauliPrinciple commented on sympy/tensor/tensor.py:
```diff
@@ -38,6 +38,7 @@
 from collections import defaultdict
 import operator
 import itertools
+from sympy.core import Wild, WildFunction
```
Ideally we import from the exact module within the library (this is not done everywhere in SymPy because this was not done so in the past).

Also keep the newline between the external and internal imports.

Kishore96in replied:
I've changed it.

ThePauliPrinciple commented on sympy/tensor/tensor.py:
```diff
@@ -2385,6 +2386,32 @@ def _expand_partial_derivative(self):
+        #Since canon_bp may be expensive, we first check for exact equality.
+        diff = self - expr
+        if (diff == 0 or
+            diff.canon_bp() == 0
```
Wouldn't this only ever prevent `S.Zero.canon_bp()` from being evaluated? Is that slow in some cases?

The SymPy object for `0` is `S.Zero`

Kishore96in replied:
I've changed it now since both take under a millisecond.

```
In [8]: %timeit S.Zero == S.Zero
353 ns ± 9.54 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [9]: %timeit canon_bp(S.Zero) == S.Zero
683 ns ± 10.2 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

Kishore96in committed:
tensor: import Wild and WildFunction from correct module

Kishore96in committed:
TensExpr._matches_simple: use S.Zero rather than 0

Kishore96in committed:
TensExpr._matches_simple: directly use canon_bp

Both take under a millisecond, so I guess there is no need to optimize
that.

In [8]: %timeit S.Zero == S.Zero
353 ns ± 9.54 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops
each)
In [9]: %timeit canon_bp(S.Zero) == S.Zero
683 ns ± 10.2 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops
each)

Kishore96in committed:
Tensor._get_symmetrized_forms: fix code quality

Pointed out by flake8-comprehensions

Kishore96in committed:
TensMul._matches_commutative: fix code quality

Pointed out by flake8-comprehensions

Kishore96in committed:
TensMul._matches_commutative: fix code quality

Pointed out by flake8-comprehensions

Kishore96in committed:
TensMul.matches: fix code quality

Pointed out by flake8-comprehensions

Kishore96in committed:
test_tensor/test_tensor_matching: fix code quality

Pointed out by flake8-comprehensions

Kishore96in committed:
TensMul._matches_commutative: simplify an expression

Kishore96in committed:
TensMul._matches_commutative: fix an edge case involving automatic index relabelling

Test coming in next commit

It sometimes happens that dummy indices are relabelled during the
recursive matching process, leading to wrong results (and subsequent
errors due to conflicting indices).

Kishore96in committed:
test_tensor: add test for an edge case involving automatic index relabelling

Kishore96in commented:
I am splitting this into smaller PRs now. The first of the series is #26335
